pymc3/variational/opvi.py

```diff
@@ -0,0 +1,488 @@
+            self.test_params = tf.params
+            self.obj_params = op.approx.params
+
+        def __call__(self, z):
+            return self.obj(z)
+
+        def updates(self, z, obj_optimizer=None, test_optimizer=None, more_params=None):
+            if more_params is None:
+                more_params = []
+            if obj_optimizer is None:
+                obj_optimizer = adagrad_optimizer(learning_rate=.001, epsilon=.1)
+            if test_optimizer is None:
+                obj_optimizer = adagrad_optimizer(learning_rate=.001, epsilon=.1)
+            target = self(z)
+            updates = theano.OrderedUpdates()
+            updates.update(obj_optimizer(target, self.obj_params + more_params))
+            if self.test_params:
+                updates.update(test_optimizer(-target, self.test_params))
+            else:
+                pass
+            return updates
+
+
+def cast_to_list(params):
+    if isinstance(params, list):
+        return params
+    elif isinstance(params, dict):
+        return params.values()
+    elif isinstance(params, theano.compile.SharedVariable):
+        return [params]
+    elif params is None:
+        return []
+    else:
+        raise TypeError('Unknown type %s for %r, need list, dict or shared variable')
+
+
+class TestFunction(object):
+    shared_params = None
+
+    @property
+    def params(self):
+        return cast_to_list(self.shared_params)
+
+    def __call__(self, z):
+        raise NotImplementedError
+
+    @classmethod
+    def from_function(cls, f):
+        if not callable(f):
+            raise ValueError('Need callable, got %r' % f)
+        obj = TestFunction()
+        obj.__call__ = f
+        return obj
+
+
+class Approximation(object):
+    def __init__(self, local_rv=None, model=None):
+        model = modelcontext(model)
+        self.model = model
+        self.check_model(model)
+        if local_rv is None:
+            local_rv = {}
+
+        def get_transformed(v):
+            if hasattr(v, 'transformed'):
+                return v.transformed
+            return v
+
+        known = {get_transformed(k): v for k, v in local_rv.items()}
+        self.known = known
+        self.local_vars = [v for v in model.free_RVs if v in known]
+        self.global_vars = [v for v in model.free_RVs if v not in known]
+        self.order, self.flat_view, self._view = flatten_model(
+            model=self.model,
+            vars=self.local_vars + self.global_vars
+        )
+        self.shared_params = self.create_shared_params()
+
+    input = property(lambda self: self.flat_view.input)
+
+    @staticmethod
+    def check_model(model):
+        """
+        Checks that model is valid for variational inference
+        """
+        vars_ = [var for var in model.vars if not isinstance(var, pm.model.ObservedRV)]
+        if any([var.dtype in pm.discrete_types for var in vars_]):
+            raise ValueError('Model should not include discrete RVs')
```

Comment:
In the paper authors suggest the way how to deal with discrete vars. The approach is described in the appendix. See https://arxiv.org/abs/1610.09033